components/sup/src/manager/file_watcher.rs

```diff
@@ -28,1215 +26,296 @@ use std::{collections::{hash_map::Entry,
-    fn indented_to_string(&self, spaces: &str, repeat: usize) -> String {
-        let mut formatter = IndentedStructFormatter::new("CommonGenerator", spaces, repeat);
-        formatter.add("prev", &self.prev);
-        formatter.add("old_prev", &self.old_prev);
-        formatter.add("path", &self.path);
-        formatter.add("split_path", &self.split_path);
-        formatter.add_string("index", self.index.to_string());
-        formatter.add_debug("path_rest", &self.path_rest);
-        formatter.fmt()
-    }
-}
-
-// An item we are interested in.
-enum WatchedFile {
-    Regular(Common),
-    MissingRegular(Common),
-    Symlink(Common),
-    Directory(Common),
-    MissingDirectory(Common),
+#[derive(Clone, Debug)]
+/// WatchedFile bundles data together concerning the file that is to be watched.  
+struct WatchedFile {
+    /// The file that is to be watched.
+    path:               PathBuf,
+    /// The directory containing the file that we are watching.
+    directory:          PathBuf,
+    /// The last observed state of the file being being watched
+    state:              WatchedFileState,
+    /// Used to signal if an initial Callbacks::file_appeared should be sent
+    /// when creating the watched file.  The file must exist on the file system
+    /// and this must be set true.  After the first execution this value will
+    /// set back to false as the intial true value has fulfilled its purpose.
+    send_initial_event: bool,
 }
 
 impl WatchedFile {
-    fn get_common(&self) -> &Common {
-        match self {
-            WatchedFile::Regular(ref c)
-            | WatchedFile::MissingRegular(ref c)
-            | WatchedFile::Symlink(ref c)
-            | WatchedFile::Directory(ref c)
-            | WatchedFile::MissingDirectory(ref c) => c,
+    fn new<P>(path: P, send_initial_event: bool) -> Result<Self>
+        where P: AsRef<Path> + Debug
+    {
+        trace!("WatchedFile::new({:?},{:?})", path, send_initial_event);
+        let path = path.as_ref();
+        Self::preclude_directories(path)?;
+        let directory = path.parent().map(PathBuf::from).unwrap();
+        let state = Self::assess_state_of(path, &directory);
+        let path = path.to_path_buf();
+        Ok(Self { path,
+                  directory,
+                  send_initial_event,
+                  state })
+    }
+
+    /// This method is used to preclude paths that we can reasonably say ares directories as
+    /// opposed to files. Necessary because we watch files, not directories, but we also need to
+    /// watch for files that may not exist so the use of the hueristics herein.
+    fn preclude_directories(path: &Path) -> Result<()> {
+        trace!("WatchedFile::preclude_directories({:?})", path);
+        if path.is_dir() {
+            let s = format!("{} is a directory instead of a file.", path.display());
+            Err(Error::FileNotFound(s))
+        } else if Self::ends_with_a_path_separator(path) {
```

Comment:
I thought that made sense too, and said so when we discussed this, but what I had forgotten until I went to try the ends_with solution is that ends_with function "only considers whole path components to match" which means that its not going to be able to zero in on the separator because its using the separator to break apart the things it will match on.  Having hit that again caused me to recall that's why I went with the implementation that I did in the first place. Giving the requirement that we are watching files which may come into and go out of existence I think there's value in sniffing out what are clearly directory paths when we can so I'm leaving this for now.

Reply:
Then how about path.to_string_lossy().ends_with(std::path::MAIN_SEPARATOR.to_string())?

Reply:
I've implemented this but kept the enclosing method for more targeted testing as we discussed.